packages/apollo-server-memcached/src/index.ts

```diff
@@ -2,31 +2,109 @@ import { KeyValueCache } from 'apollo-server-caching';
 import Memcached from 'memcached';
 import { promisify } from 'util';
 
+/**
+ * Defines a KeyValueCache implementation that supports cache tag invalidation.
+ * When cache entry is associated with 1 or more tags, we store tag:version
+ * pairs as metadata, as well as the current version associated with each tag.
+ * Version numbers are simply integers that are incremented whenever a tag is
+ * invalidated. For a cache entry to be valid, all of its associated tags must
+ * have version numbers that match their current version numbers.
+ *
+ * Performance:
+ * - set()        : # of db reads = # of tags
+ * - get()        : # of db reads = # of tags
+ * - invalidate() : # of db writes/incr = 1 for each tag
+ *
+ * Storage overhead:
+ * - key->data + tag:version pairs for each tag
+ * - tag->version stored for each tag
+ */
 export default class MemcachedKeyValueCache implements KeyValueCache {
   readonly client;
   readonly defaultSetOptions = {
     ttl: 300,
+    tags: [],
   };
 
   constructor(serverLocation: Memcached.Location, options?: Memcached.options) {
     this.client = new Memcached(serverLocation, options);
     // promisify client calls for convenience
     this.client.get = promisify(this.client.get).bind(this.client);
+    this.client.getMulti = promisify(this.client.getMulti).bind(this.client);
+    this.client.incr = promisify(this.client.incr).bind(this.client);
     this.client.set = promisify(this.client.set).bind(this.client);
     this.client.flush = promisify(this.client.flush).bind(this.client);
   }
 
   async set(
     key: string,
     data: string,
-    options?: { ttl?: number },
+    options?: { ttl?: number; tags?: string[] },
   ): Promise<void> {
-    const { ttl } = Object.assign({}, this.defaultSetOptions, options);
-    await this.client.set(key, data, ttl);
+    const { ttl, tags } = Object.assign({}, this.defaultSetOptions, options);
+
+    // get current versions for all tags
+    let tagVersions = {};
+    if (tags.length > 0) {
+      tagVersions = await this.client.getMulti(tags);
+
+      // initialize tag versions that don't exist yet
+      const setOperations: any[] = [];
```

Comment:
Although I also think `set` is potentially unsafe here, because another process may have modified the tag in the meantime (i.e. after reading) and setting it to 1 would interfere with ongoing invalidations. Is there any way we could use an 'increment with intitial' here? Or else maybe an `add' followed by an `incr`?

Reply:
```typescript
      const unknownTags = tags.filter(tag => !currentTagVersions[tag]);
      for (const version of await Promise.all(
        unknownTags.map(tag => {
          this.client.add(tag);
          return this.client.incr(tag, 1);
        }),
      )) {
        currentTagVersions[version];
      }
```

Reply:
Nice way to deal with safety issues!